File: server/src/main/com/broll/gainea/server/core/goals/OccupyGoal.java
package com.broll.gainea.server.core.goals;

import com.broll.gainea.server.core.bot.strategy.GoalStrategy;
import com.broll.gainea.server.core.map.Area;
import com.broll.gainea.server.core.map.AreaCollection;
import com.broll.gainea.server.core.map.AreaID;
import com.broll.gainea.server.core.map.ContinentID;
import com.broll.gainea.server.core.map.Expansion;
import com.broll.gainea.server.core.map.ExpansionType;
import com.broll.gainea.server.core.map.IslandID;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.map.MapContainer;
import com.broll.gainea.server.core.objects.MapObject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class OccupyGoal extends Goal {

    private final static Logger Log = LoggerFactory.getLogger(OccupyGoal.class);
    private List<Location> locations = new ArrayList<>();
    private Map<Location, Function<Location, Boolean>> conditions = new HashMap<>();
    protected MapContainer map;

    public OccupyGoal(GoalDifficulty difficulty, String text) {
        super(difficulty, text);
    }

    @Override
    public void botStrategy(GoalStrategy strategy) {
        strategy.setRequiredUnits(locations.size());
        strategy.updateTargets(locations.stream().collect(Collectors.toSet()));
    }

    @Override
    protected boolean validForGame() {
        map = game.getMap();
        try {
            initOccupations();
        } catch (MissingExpansionException e) {
            return false;
        }
        //if any area is null, then its not part of the expansions, so goal is invalid
        for (Location area : locations) {
            if (area == null) {
                return false;
            }
        }
        if (locations.isEmpty()) {
            //no locations -> missing expansion
            return false;
        }
        //set expansion restrictions by required locations
        setExpansionRestriction(Arrays.stream(ExpansionType.values()).filter(this::hasLocationsOf).toArray(ExpansionType[]::new));
        setProgressionGoal(locations.size());
        return true;
    }

    private boolean hasLocationsOf(ExpansionType type) {
        for (Location location : locations) {
            Expansion expansion = map.getExpansion(type);
            if (expansion != null && expansion.getAllLocations().contains(location)) {
                return true;
            }
        }
        return false;
    }

    protected abstract void initOccupations();

    private void assureIslandExists(IslandID... islands) {
        for (IslandID island : islands) {
            if (map.getIsland(island) == null) {
                throw new MissingExpansionException();
            }
        }
    }

    private void assureContinentExists(ContinentID... continents) {
        for (ContinentID continent : continents) {
            if (map.getContinent(continent) == null) {
                throw new MissingExpansionException();
            }
        }
    }

    private void assureExpansionExists(ExpansionType... expansions) {
        for (ExpansionType expansion : expansions) {
            if (map.getExpansion(expansion) == null) {
                throw new MissingExpansionException();
            }
        }
    }

    protected List<Area> occupy(AreaID... areas) {
        List<Area> list = Arrays.stream(areas).map(map::getArea).filter(Objects::nonNull).collect(Collectors.toList());
        this.locations.addAll(list);
        return list;
    }

    protected List<Area> occupy(Function<Area, Boolean> filter, AreaID... areas) {
        List<Area> list = Arrays.stream(areas).map(map::getArea).filter(Objects::nonNull).collect(Collectors.toList());
        list.removeIf(it -> !filter.apply(it));
        this.locations.addAll(list);
        return list;
    }

    protected List<Area> occupy(IslandID... islands) {
        assureIslandExists(islands);
        List<Area> list = new ArrayList<>();
        Arrays.stream(islands).map(map::getIsland).filter(Objects::nonNull).map(AreaCollection::getAreas).forEach(areas -> list.addAll(areas));
        locations.addAll(list);
        return list;
    }

    protected List<Area> occupy(Function<Area, Boolean> filter, IslandID... islands) {
        assureIslandExists(islands);
        List<Area> list = new ArrayList<>();
        Arrays.stream(islands).map(map::getIsland).filter(Objects::nonNull).map(AreaCollection::getAreas).forEach(areas -> list.addAll(areas));
        list.removeIf(it -> !filter.apply(it));
        locations.addAll(list);
        return list;
    }

    protected List<Area> occupy(ContinentID... continents) {
        assureContinentExists(continents);
        List<Area> list = new ArrayList<>();
        Arrays.stream(continents).map(map::getContinent).filter(Objects::nonNull).map(AreaCollection::getAreas).forEach(areas -> list.addAll(areas));
        locations.addAll(list);
        return list;
    }

    protected List<Area> occupy(Function<Area, Boolean> filter, ContinentID... continents) {
        assureContinentExists(continents);
        List<Area> list = new ArrayList<>();
        Arrays.stream(continents).map(map::getContinent).filter(Objects::nonNull).map(AreaCollection::getAreas).forEach(areas -> list.addAll(areas));
        list.removeIf(it -> !filter.apply(it));
        locations.addAll(list);
        return list;
    }

    protected List<Area> occupy(ExpansionType... expansions) {
        assureExpansionExists(expansions);
        List<Area> list = new ArrayList<>();
        Arrays.stream(expansions).map(map::getExpansion).filter(Objects::nonNull).map(Expansion::getContents).forEach(col -> col.stream().map(AreaCollection::getAreas).forEach(areas -> list.addAll(areas)));
        locations.addAll(list);
        return list;
    }

    protected List<Area> occupy(Function<Area, Boolean> filter, ExpansionType... expansions) {
        assureExpansionExists(expansions);
        List<Area> list = new ArrayList<>();
        Arrays.stream(expansions).map(map::getExpansion).filter(Objects::nonNull).map(Expansion::getContents).forEach(col -> col.stream().map(AreaCollection::getAreas).forEach(areas -> list.addAll(areas)));
        list.removeIf(it -> !filter.apply(it));
        locations.addAll(list);
        return list;
    }

    protected List<Location> occupy(List<Location> locations) {
        this.locations.addAll(locations);
        return locations;
    }

    protected List<Location> occupy(Location... locations) {
        this.locations.addAll(Arrays.asList(locations));
        return Arrays.asList(locations);
    }

    protected List<Location> occupy(Stream<? extends Location> stream) {
        List<Location> list = stream.collect(Collectors.toList());
        this.locations.addAll(list);
        return list;
    }

    protected void condition(List<Location> locations, Function<Location, Boolean>... conditions) {
        locations.forEach(location -> this.conditions.put(location, loc ->
                Arrays.stream(conditions).map(it -> it.apply(loc)).reduce(true, Boolean::logicalAnd)
        ));
    }

    @Override
    public void check() {
        List<Location> occupiedLocations = player.getControlledLocations();
        boolean success = true;
        int progress = 0;
        for (Location location : locations) {
            Function<Location, Boolean> condition = conditions.get(location);
            if (condition != null) {
                if (condition.apply(location)) {
                    progress++;
                } else {
                    //condition for the location not satisfied
                    success = false;
                }
            } else {
                //default condition: simply occupied
                if (occupiedLocations.contains(location)) {
                    progress++;
                } else {
                    //area not occupied by player
                    success = false;
                }
            }
        }
        updateProgression(progress);
        if (success) {
            success();
        }
    }

    @Override
    public void moved(List<MapObject> units, Location location) {
        if (units.get(0).getOwner() == player) {
            //unit of this player moved, check occupy condition
            check();
        }
    }

    @Override
    public void spawned(MapObject object, Location location) {
        if (object.getOwner() == player) {
            //unit of this player spawned, check occupy condition
            check();
        }
    }
}

File: server/src/main/com/broll/gainea/server/core/goals/OccupyGoal.kt
package com.broll.gainea.server.core.goals

import com.broll.gainea.server.core.bot.strategy.GoalStrategy
import com.broll.gainea.server.core.map.Area
import com.broll.gainea.server.core.map.AreaID
import com.broll.gainea.server.core.map.ContinentID
import com.broll.gainea.server.core.map.ExpansionType
import com.broll.gainea.server.core.map.IslandID
import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.map.MapContainer
import com.broll.gainea.server.core.objects.MapObject
import org.slf4j.LoggerFactory

abstract class OccupyGoal(difficulty: GoalDifficulty, text: String) : Goal(difficulty, text) {
    private val conditions = HashMap<Location, (Location) -> Boolean>()
    protected lateinit var map: MapContainer
    protected var autoCheckProgressions = true
    override fun botStrategy(strategy: GoalStrategy) {
        strategy.setRequiredUnits(locations.size)
        strategy.updateTargets(locations.toHashSet())
    }

    override fun validForGame(): Boolean {
        map = game.map
        try {
            initOccupations()
        } catch (e: MissingExpansionException) {
            return false
        }
        if (locations.isEmpty()) {
            //no locations -> missing expansion
            return false
        }
        //set expansion restrictions by required locations
        setExpansionRestriction(*ExpansionType.entries.filter { hasLocationsOf(it) }.toTypedArray())
        if (autoCheckProgressions) {
            progressionGoal = locations.size
        }
        return true
    }

    private fun hasLocationsOf(type: ExpansionType): Boolean {
        for (location in locations) {
            val expansion = map.getExpansion(type)
            if (expansion != null && expansion.allLocations.contains(location)) {
                return true
            }
        }
        return false
    }

    protected abstract fun initOccupations()

    private fun assureAreaExists(vararg areas: AreaID) {
        if (areas.any { map.getArea(it) == null }) {
            throw MissingExpansionException()
        }
    }

    private fun assureIslandExists(vararg islands: IslandID) {
        if (islands.any { map.getIsland(it) == null }) {
            throw MissingExpansionException()
        }
    }

    private fun assureContinentExists(vararg continents: ContinentID) {
        if (continents.any { map.getContinent(it) == null }) {
            throw MissingExpansionException()
        }
    }

    private fun assureExpansionExists(vararg expansions: ExpansionType) {
        if (expansions.any { map.getExpansion(it) == null }) {
            throw MissingExpansionException()
        }
    }

    protected fun occupy(vararg areas: AreaID): List<Area> {
        assureAreaExists(*areas)
        val list = areas.mapNotNull { map.getArea(it) }
        locations.addAll(list)
        return list
    }

    protected fun occupy(filter: (Area) -> Boolean, vararg areas: AreaID): List<Area> {
        assureAreaExists(*areas)
        val list = areas.mapNotNull { map.getArea(it) }.filter(filter)
        locations.addAll(list)
        return list
    }

    protected fun occupy(vararg islands: IslandID): List<Area> {
        assureIslandExists(*islands)
        val list = islands.mapNotNull { map.getIsland(it) }.flatMap { it.areas }
        locations.addAll(list)
        return list
    }

    protected fun occupy(filter: (Area) -> Boolean, vararg islands: IslandID): List<Area> {
        assureIslandExists(*islands)
        val list = islands.mapNotNull { map.getIsland(it) }.flatMap { it.areas }.filter(filter)
        locations.addAll(list)
        return list
    }

    protected fun occupy(vararg continents: ContinentID): List<Area> {
        assureContinentExists(*continents)
        val list = continents.mapNotNull { map.getContinent(it) }.flatMap { it.areas }
        locations.addAll(list)
        return list
    }

    protected fun occupy(filter: (Area) -> Boolean, vararg continents: ContinentID): List<Area> {
        assureContinentExists(*continents)
        val list =
            continents.mapNotNull { map.getContinent(it) }.flatMap { it.areas }.filter(filter)
        locations.addAll(list)
        return list
    }

    protected fun occupy(vararg expansions: ExpansionType): List<Area> {
        assureExpansionExists(*expansions)
        val list = expansions.mapNotNull { map.getExpansion(it) }.flatMap { it.allAreas }
        locations.addAll(list)
        return list
    }

    protected fun occupy(filter: (Area) -> Boolean, vararg expansions: ExpansionType): List<Area> {
        assureExpansionExists(*expansions)
        val list =
            expansions.mapNotNull { map.getExpansion(it) }.flatMap { it.allAreas }.filter(filter)
        locations.addAll(list)
        return list
    }

    protected fun occupy(locations: List<Location>): List<Location> {
        this.locations.addAll(locations.toList())
        return locations.toList()
    }

    protected fun occupy(vararg locations: Location): List<Location> {
        this.locations.addAll(locations.toList())
        return locations.toList()
    }

    protected fun condition(locations: List<Location>, vararg conditions: (Location) -> Boolean) {
        locations.forEach {
            this.conditions[it] = { location ->
                conditions.all { condition -> condition(location) }
            }
        }
    }

    override fun check() {
        val occupiedLocations = player.controlledLocations
        var success = true
        var progress = 0
        for (location in locations) {
            val condition = conditions[location]
            if (condition != null) {
                if (condition(location)) {
                    progress++
                } else {
                    //condition for the location not satisfied
                    success = false
                }
            } else {
                //default condition: simply occupied
                if (occupiedLocations.contains(location)) {
                    progress++
                } else {
                    //area not occupied by player
                    success = false
                }
            }
        }
        if (autoCheckProgressions) {
            updateProgression(progress)
        }
        if (success) {
            success()
        }
    }

    override fun unitsMoved(units: List<MapObject>, location: Location) {
        if (units.any { it.owner === player }) {
            //unit of this player moved, check occupy condition
            check()
        }
    }

    override fun unitSpawned(`object`: MapObject, location: Location) {
        if (`object`.owner === player) {
            //unit of this player spawned, check occupy condition
            check()
        }
    }

    companion object {
        private val Log = LoggerFactory.getLogger(OccupyGoal::class.java)
    }
}
